pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/creator/IndexCreatorProvider.java

```diff
@@ -0,0 +1,28 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.spi.creator;
+
+/**
+ * Plugin interface to abstract index creation.
+ */
+public interface IndexCreatorProvider
```

Comment:
Should we consider putting all methods into this single interface instead of splitting them into providers for each index type since this is the pluggable unit?

Reply:
I prefer it this way because it allows the code in pinot-segment-local to request as little functionality as it needs - in order to construct a forward index I shouldn't need the power to create a bloom filter. These types can be composed as generic intersection types when more than one service is required. In the future, we may want to change the implementations used within core, and shouldn't need to implement bloom filter provision to construct an object which needs to create forward indexes. The fact that there is only one override at the SPI level is an implementation detail of the override mechanism, and pinot-segment-local/pinot-core should not be aware of this.